File: backend/scripts/create_profiles_from_uploads.js
const fs = require('fs');
const path = require('path');
const { dbAll, dbGet, dbRun } = require('../../Database/db');
const { ALL_SUBJECTS } = require('../models/teacherModel');

function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function randomElement(arr) {
  return arr[randomInt(0, arr.length - 1)];
}

function generatePhone() {
  // Formato argentino aproximado: +54 9 11 1234 5678
  const area = randomElement(['11','221','351','341','297']);
  const part1 = String(randomInt(1000, 9999));
  const part2 = String(randomInt(1000, 9999));
  return `+54 9 ${area} ${part1} ${part2}`;
}

async function ensureModalitiesColumn() {
  const cols = await dbAll("PRAGMA table_info(teachers)");
  const has = cols.some(c => c.name === 'modalities');
  if (!has) {
    console.log('Añadiendo columna `modalities` a la tabla teachers...');
    await dbRun('ALTER TABLE teachers ADD COLUMN modalities TEXT');
    console.log('Columna `modalities` creada');
  } else {
    console.log('Columna `modalities` ya existe');
  }
}

async function run() {
  try {
    await ensureModalitiesColumn();

    const uploadsRoot = path.join(__dirname, '../../Frontend/Assets/uploads');

    // Definir listas de nombres por género para evitar mezclar nombres/sexo
    const genders = [
      {
        dir: 'hombres',
        firstNames: ['Juan', 'Carlos', 'Diego', 'Miguel', 'Jorge', 'Luis', 'Antonio', 'Marco', 'Ricardo', 'Javier', 'Samuel', 'Alejandro', 'Andrés', 'Guillermo', 'Felipe', 'Matías']
      },
      {
        dir: 'mujeres',
        firstNames: ['María', 'Ana', 'Laura', 'Isabel', 'Patricia', 'Sofía', 'Elena', 'Gabriela', 'Rosa', 'Lucia', 'Martina', 'Valeria', 'Carolina', 'Cristina', 'Adriana', 'Camila']
      }
    ];

    const lastNames = ['García','Pérez','Rodríguez','Gómez','López','Martínez','Santos','Silva','Rojas','Fernández','Vargas','Reyes','Fuentes','Cortés','Medina','Rojas'];

    // Small curriculum and schedules pools
    const curriculums = [
      'Programa estructurado de teoría y práctica',
      'Clases personalizadas con ejercicios y seguimiento',
      'Temario orientado a resolución de problemas y exámenes',
      'Enfoque práctico con proyectos y evaluación continua'
    ];

    const schedules = [
      'Lun/Mié/Vie 18:00-20:00',
      'Mar/Jue 10:00-12:00',
      'Sábados 09:00-12:00',
      'Horario flexible'
    ];

    const locations = ['CABA - Centro','CABA - Norte','Gran Buenos Aires','La Plata'];

    // Limpiar tabla teachers para crear perfiles desde las fotos (asegurar único set)
    console.log('Eliminando profesores existentes (si los hay)...');
    await dbRun('DELETE FROM teachers');
    // Resetear AUTOINCREMENT (si existe sqlite_sequence)
    try {
      await dbRun('DELETE FROM sqlite_sequence WHERE name = ?', ['teachers']);
    } catch (err) {
      // ignora si sqlite_sequence no existe
    }

    for (const g of genders) {
      const gender = g.dir;
      const dir = path.join(uploadsRoot, gender);
      if (!fs.existsSync(dir)) continue;

      const files = fs.readdirSync(dir).filter(f => /\.(jpe?g|png)$/i.test(f));
      for (const file of files) {
        const originalPath = path.join(dir, file);

        // Compose teacher data
        const firstName = randomElement(g.firstNames);
        const lastName = randomElement(lastNames);
        const age = randomInt(25, 60);
        const email = `teacher_${gender}_${Date.now()}_${randomInt(1000,9999)}@example.com`;
        const phone = generatePhone();
        const subjectsCount = randomInt(2,4);
        const subjects = [...ALL_SUBJECTS].sort(() => 0.5 - Math.random()).slice(0, subjectsCount);

        // Modalities: 20% both, else single
        let modalities = [];
        if (Math.random() < 0.2) {
          modalities = ['virtual','presencial'];
        } else {
          modalities = [randomElement(['virtual','presencial'])];
        }

        const modality = modalities[0];

        const curriculum = randomElement(curriculums);
        const schedule = randomElement(schedules);
        const location = modalities.includes('presencial') ? randomElement(locations) : null;
        // No incluir información de contacto en la descripción
        const description = `Profesor/a con experiencia en ${subjects.join(', ')}.`;

        // Insert into DB
        const sql = `INSERT INTO teachers (
          firstName, lastName, age, email, phone, description, curriculum,
          photo, classSize, subjects, modality, modalities, schedules, location, views
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`;

        const photoRelPath = `/assets/uploads/${gender}/${file}`; // se actualizará tras renombrar
        const params = [
          firstName, lastName, age, email, phone, description, curriculum,
          photoRelPath, randomInt(1,5), JSON.stringify(subjects), modality, JSON.stringify(modalities), schedule, location, randomInt(0,50)
        ];

        const res = await dbRun(sql, params);
        const id = res.id;

        // Rename file to use ID
        const ext = path.extname(file).toLowerCase();
        const newFilename = `${id}${ext}`;
        const newPath = path.join(dir, newFilename);
        try {
          if (originalPath !== newPath) fs.renameSync(originalPath, newPath);
        } catch (err) {
          console.warn('No se pudo renombrar', originalPath, err.message);
        }

        const newPhotoRel = `/assets/uploads/${gender}/${newFilename}`;
        await dbRun('UPDATE teachers SET photo = ? WHERE id = ?', [newPhotoRel, id]);

        console.log(`Creado profesor ID=${id} (${firstName} ${lastName}) -> ${newPhotoRel}`);
      }
    }

    console.log('\nTodos los perfiles creados/actualizados a partir de las fotos.');
  } catch (err) {
    console.error('Error creando perfiles:', err);
  } finally {
    process.exit(0);
  }
}

run();

File: Backend/scripts/create_profiles_from_uploads.py
#!/usr/bin/env python3
import os
import sqlite3
import json
import random
import time
import re
import shutil
import sys


def random_int(a, b):
    return random.randint(a, b)


def random_element(seq):
    return random.choice(seq)


def generate_phone():
    # Formato argentino aproximado: +54 9 (area) xxxx xxxx
    area = random_element(['11', '221', '351', '341', '297'])
    part1 = str(random_int(1000, 9999))
    part2 = str(random_int(1000, 9999))
    return f"+54 9 {area} {part1} {part2}"


def parse_all_subjects(model_path):
    try:
        text = open(model_path, 'r', encoding='utf-8').read()
        m = re.search(r"const\s+ALL_SUBJECTS\s*=\s*\[([\s\S]*?)\];", text)
        if not m:
            return None
        inner = m.group(1)
        # Find all quoted strings
        items = re.findall(r"'(.*?)'|\"(.*?)\"", inner)
        subjects = [a if a else b for (a, b) in items]
        return subjects
    except Exception:
        return None


def ensure_modalities_column(conn):
    cur = conn.cursor()
    cur.execute("PRAGMA table_info(teachers)")
    cols = [row[1] for row in cur.fetchall()]
    if 'modalities' not in cols:
        print('Añadiendo columna `modalities` a la tabla teachers...')
        cur.execute('ALTER TABLE teachers ADD COLUMN modalities TEXT')
        conn.commit()
        print('Columna `modalities` creada')
    else:
        print('Columna `modalities` ya existe')


def main():
    base = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))
    db_path = os.path.join(base, 'Database', 'estudius.db')
    schema_path = os.path.join(base, 'Database', 'schema.sql')
    model_path = os.path.join(base, 'Backend', 'models', 'teacherModel.js')
    uploads_root = os.path.join(base, 'Frontend', 'Assets', 'uploads')

    if not os.path.exists(db_path):
        print('Base de datos no encontrada en', db_path)
        # Create empty DB file; schema will be applied

    # Load subjects list from model to keep consistency with frontend
    subjects = parse_all_subjects(model_path) or [
        'Matemática', 'Lengua', 'Historia', 'Programación', 'Inglés', 'Física', 'Química'
    ]

    conn = sqlite3.connect(db_path)
    cur = conn.cursor()

    # Ensure schema is applied (executes CREATE TABLE IF NOT EXISTS)
    try:
        if os.path.exists(schema_path):
            schema = open(schema_path, 'r', encoding='utf-8').read()
            cur.executescript(schema)
            conn.commit()
    except Exception as e:
        print('Aviso: no pude aplicar schema.sql:', e)

    ensure_modalities_column(conn)

    genders = ['hombres', 'mujeres']

    first_names = ['Alex','Mateo','Lucas','Santiago','Martín','Diego','Lucía','María','Valentina','Sofía','Camila','Ana','Paula','Julia']
    last_names = ['García','Pérez','Rodríguez','Gómez','López','Martínez','Santos','Silva','Rojas','Fernández']
    curriculums = [
        'Programa estructurado de teoría y práctica',
        'Clases personalizadas con ejercicios y seguimiento',
        'Temario orientado a resolución de problemas y exámenes',
        'Enfoque práctico con proyectos y evaluación continua'
    ]
    schedules = [
        'Lun/Mié/Vie 18:00-20:00',
        'Mar/Jue 10:00-12:00',
        'Sábados 09:00-12:00',
        'Horario flexible'
    ]
    locations = ['CABA - Centro','CABA - Norte','Gran Buenos Aires','La Plata']

    total_created = 0
    for gender in genders:
        dir_path = os.path.join(uploads_root, gender)
        if not os.path.isdir(dir_path):
            print('No existe carpeta:', dir_path)
            continue

        files = [f for f in os.listdir(dir_path) if os.path.splitext(f)[1].lower() in ('.jpg', '.jpeg', '.png')]
        for fname in files:
            original = os.path.join(dir_path, fname)
            try:
                first = random_element(first_names)
                last = random_element(last_names)
                age = random_int(25, 60)
                email = f"teacher_{int(time.time()*1000)}_{random_int(1000,9999)}@example.com"
                phone = generate_phone()
                subj_count = random_int(2, min(4, max(2, len(subjects))))
                subj = random.sample(subjects, subj_count)

                if random.random() < 0.2:
                    modalities = ['virtual', 'presencial']
                else:
                    modalities = [random_element(['virtual', 'presencial'])]
                modality = modalities[0]

                curriculum = random_element(curriculums)
                schedule = random_element(schedules)
                location = random_element(locations) if 'presencial' in modalities else None
                description = f"Profesor/a con experiencia en {', '.join(subj)}."

                photo_rel = f"/assets/uploads/{gender}/{fname}"
                class_size = random_int(1, 5)
                views = random_int(0, 50)

                sql = (
                    "INSERT INTO teachers (firstName, lastName, age, email, phone, description, curriculum, "
                    "photo, classSize, subjects, modality, modalities, schedules, location, views) "
                    "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
                )
                params = [
                    first, last, age, email, phone, description, curriculum,
                    photo_rel, class_size, json.dumps(subj), modality, json.dumps(modalities), schedule, location, views
                ]

                cur.execute(sql, params)
                tid = cur.lastrowid
                conn.commit()

                # Rename file to <id>.<ext>
                ext = os.path.splitext(fname)[1].lower()
                new_fname = f"{tid}{ext}"
                new_path = os.path.join(dir_path, new_fname)
                shutil.move(original, new_path)

                new_photo_rel = f"/assets/uploads/{gender}/{new_fname}"
                cur.execute('UPDATE teachers SET photo = ? WHERE id = ?', (new_photo_rel, tid))
                conn.commit()

                print(f"Creado profesor ID={tid} -> {new_photo_rel}")
                total_created += 1
            except Exception as e:
                print('Error con archivo', fname, e)

    conn.close()
    print('\nTotal perfiles creados/actualizados:', total_created)


if __name__ == '__main__':
    main()
